### PR TITLE
feat: infer mutual structural recursion

### PR DESCRIPTION
**Pull request by nomeata**
the support for mutual structural recursion (new since #4575) is extended so that Lean tries to infer it even without annotations.

* The error message when termination checking fails looks quite different now. Maybe a bit better, maybe with more room for improvements.
* If there are too many combinations (with an arbitrary cut-off) for a given argument type, it will just give up and ask the user to use `termination_by structural`.
* It is now legal to specify `termination_by structural` on not necessarily all functions of a clique; this simply restricts the combinations of arguments that Lean considers.

### CODE REVIEW

File: tests/lean/run/structuralMutual.lean [hunk 1]
```diff
@@ -1,5 +1,13 @@
 import Lean.Elab.Command
 
+/-!
+Mutual structural recursion.
+
+In this file we often attach a `termination_By structural` annotation to at least
+one of the functions to force strucural recursion. We don't want lean to resort to
```
Comment (tobiasgrosser) [inline tests/lean/run/structuralMutual.lean:7]:
```suggestion

In this file we often attach a `termination_by structural` annotation to at least
one of the functions to force strucural recursion. We don't want lean to resort to
```

File: src/Lean/Meta/Basic.lean [hunk 2]
```diff
@@ -1492,6 +1492,16 @@ private def withLocalContextImp (lctx : LocalContext) (localInsts : LocalInstanc
 def withLCtx (lctx : LocalContext) (localInsts : LocalInstances) : n α → n α :=
   mapMetaM <| withLocalContextImp lctx localInsts
 
+/--
+Runs `k` in a local envrionment with the the `fvarIds` erased.
+-/
```
Comment (tobiasgrosser) [inline src/Lean/Meta/Basic.lean:1497]:
```suggestion
Runs `k` in a local envrionment with the `fvarIds` erased.
-/
```

File: src/Lean/Elab/PreDefinition/Structural/FindRecArg.lean [hunk 3]
```diff
@@ -85,47 +104,172 @@ def getRecArgInfo (fnName : Name) (numFixed : Nat) (xs : Array Expr) (i : Nat) :
...
+          -- not suffice?
+          let r ← k comb
+          trace[Elab.definition.structural] "tryTellArgs report:\n{report}"
+          return r
+        catch e =>
+          let m ← prettyParameterSet fnNames xs values comb
+          report := report ++ m!"Cannot use {m}:{indentD e.toMessageData}\n"
+    else
+          report := report ++ m!"Too many possible combinations of parameters of type {group} (or " ++
+            m!"please the recursive argument explicitly using `termination_by structural`).\n"
```
Comment (tobiasgrosser) [inline src/Lean/Elab/PreDefinition/Structural/FindRecArg.lean:270]: Is here a verb missing in the sentence before, e.g., please **provide** the recursive argument explicitly?

File: src/Lean/Elab/PreDefinition/Structural/FindRecArg.lean [hunk 4]
```diff
@@ -85,47 +104,172 @@ def getRecArgInfo (fnName : Name) (numFixed : Nat) (xs : Array Expr) (i : Nat) :
...
+  let mut indicesPos : HashSet Nat := {}
+  for recArgInfo in recArgInfos do
+    for pos in recArgInfo.indicesPos do
+      indicesPos := indicesPos.insert pos
+  let (indices,nonIndices) := recArgInfos.partition (indicesPos.contains ·.recArgPos)
+  return nonIndices ++ indices
+
+/--
+An instance of an mutually inductive groupa of inductives, identified by the `all` array
+and the level and expressions parameters.
```
Comment (tobiasgrosser) [inline src/Lean/Elab/PreDefinition/Structural/FindRecArg.lean:155]:
```suggestion
An instance of an mutually inductive group of inductives, identified by the `all` array
and the level and expressions parameters.
```

File: src/Lean/Elab/PreDefinition/Structural/FindRecArg.lean (hunk 3 above)
Comment (nomeata) [reply src/Lean/Elab/PreDefinition/Structural/FindRecArg.lean:270]:
```suggestion
            m!"please indicate the recursive argument explicitly using `termination_by structural`).\n"
```

**Pull request merged**